### DISCUSSION TITLE
Smart Contract Functions- AEIP17

### DISCUSSION THREAD

**Discussion by samuelmanzanera** [archethic-foundation/aeip#48] [AEIP]
This is a discussion for the AEIP17 located at: https://github.com/archethic-foundation/aeip/blob/main/AEIP-17.md

**Comment by bchamagne** [comment]
About the API, there's a lot not explained here:
- how do I have an argument that is a boolean?
- how do I have an argument that is a list ?
- how do I have an argument that is a dict ?
- how do you differentiate "123" from 123 ?
- why use query string instead of body?

For these reasons I can recommend to use the [JSON RPC 2.0 specification](https://www.jsonrpc.org/specification). It is very well known and very easy to use. It resolves the questions above by using JSON. 

Here's what would change 

```
POST /api/contract/02fa.....

REQUEST
{
   "jsonrpc": "2.0",
   "id": "123e4567-e89b-12d3-a456-426614174000",
   "method": "get_exchange_rate",
   "params": ["UCO", "BTC"]
}
SUCCESS RESULT
{
   "jsonrpc": "2.0",
   "id": "123e4567-e89b-12d3-a456-426614174000",
   "result": 0.006
}
ERROR RESULT
{
   "jsonrpc": "2.0",
   "id": "123e4567-e89b-12d3-a456-426614174000",
   "error": {"code": -32601, "message": "method not found"}
}

```

**Reply by samuelmanzanera** [reply]
I'm not against jsonrpc but if we use JSON rpc the best would have to rewrite the endpoint to remove rest calls as it might not be necessary to have multiple endpoint but a single one with several methods.

**Reply by samuelmanzanera** [reply]
But otherwise you can use the POST body.

For complex arg, we have to see if we want to support it. 
I know on Eth not all the data structure are supported as argument or return.

**Reply by samuelmanzanera** [reply]
We have to see how many argument we are limiting for the external functions .

**Reply by samuelmanzanera** [reply]
I added notion of JSON in cdbe18b15b823af135088062f53d3170d43b557a to ease the integration

**Comment by bchamagne** [comment]
Are the functions callable from the `condition` blocks ?

**Reply by samuelmanzanera** [reply]
I think so. at least the private functions.

**Reply by Neylix** [reply]
I think all functions should be callable from action and condition block. This public functions are also helpers for the contract itself.
The question is do we allow calling a function inside a function ? I think this needs to be restricted since it can create function call loop

**Reply by samuelmanzanera** [reply]
For this reason on Ethereum, there is distinction between internal or external function.
all the function cannot be used in the same scope.
Because an internal function which does a lot of things has to be payable.
While external function are only accessible by the external world and do less things

**Reply by samuelmanzanera** [reply]
What happens on Ethereum for instance, it's to decoupled internal and external functions.
For example an internal function about reading state, can be used in transaction's based function or external calls.

**Reply by Neylix** [reply]
So it means that an external function could just call an internal function ?
How can we restrict the usage of some module in external function, if they can call internal function that are not restricted ?

**Reply by samuelmanzanera** [reply]
this would be the role of interpreter parsing I guess

**Reply by samuelmanzanera** [reply]
Finally, I think it's fine to support extern function as we don't have the memory problem as other blockchains are doing.
Just the rule of external function would be the same, meaning the scope should be limited as well.
